sdk/core/System.ClientModel/src/Pipeline/ClientLoggingPolicy.cs

```diff
@@ -0,0 +1,455 @@
+        {
+            ContentTypeUtilities.TryGetTextEncoding(contentType, out requestTextEncoding);
+        }
+
+        // Log to event source
+        Singleton.RequestContent(requestId ?? string.Empty, bytes, requestTextEncoding);
+    }
+
+    private void LogResponse(PipelineResponse response, string? responseId, double elapsed)
+    {
+        bool isError = response.IsError;
+        if (isError)
+        {
+            Singleton.ErrorResponse(response, responseId ?? string.Empty, Sanitizer!, elapsed);
+        }
+        else
+        {
+            Singleton.Response(response, responseId ?? string.Empty, Sanitizer!, elapsed);
+        }
+    }
+
+    private async Task LogResponseContent(PipelineResponse response, string? responseId, bool bufferResponse, bool async, CancellationToken cancellationToken)
+    {
+        // If logging content is disabled, or the request content is null, or informational logs are not enabled, return.
+        // Checking the log level prevents expensive operations from being executed
+        // TODO - consider moving this down to LoggingStream
+        var logErrorResponseContent = response.IsError && Singleton.IsEnabled(EventLevel.Warning, EventKeywords.All);
+        var logResponseContent = Singleton.IsEnabled(EventLevel.Informational, EventKeywords.All) || logErrorResponseContent;
+
+        if (!_logContent || !logResponseContent)
+        {
+            return; // nothing to log
+        }
+
+        // Determine if the content stream should be wrapped in a logging stream, which logs the stream in blocks as it is being read
+        var wrapContent =
+            !bufferResponse // Content is not buffered - content should be in the content stream
+            && response.ContentStream != null // Content stream is available
+            && response.ContentStream?.CanSeek == false; // Content stream is not seekable - if it's seekable we can just log directly
+
+        // Try to extract a text encoding from the headers
+        Encoding? responseTextEncoding = null;
+
+        if (response.Headers.TryGetValue("Content-Type", out var contentType) && contentType != null)
+        {
+            ContentTypeUtilities.TryGetTextEncoding(contentType, out responseTextEncoding);
+        }
+
+        if (wrapContent)
+        {
+            response.ContentStream = new LoggingStream(responseId ?? string.Empty, _maxLength, response.ContentStream!, response.IsError, responseTextEncoding);
+        }
+        else
+        {
+            byte[]? bytes = null;
+            if (bufferResponse)
+            {
+                // TODO - is this the best way to do this
+                var allBytes = response.Content.ToArray();
+                var length = Math.Min(allBytes.Length, _maxLength);
+                bytes = new byte[length];
+                Buffer.BlockCopy(allBytes, 0, bytes, 0, length);
+            }
+            else
+            {
+                if (response.ContentStream != null)
```

Comment:
This comment: https://github.com/Azure/azure-sdk-for-net/pull/43510#discussion_r1575107274 still applies here:
> Is this just truncating the BinaryContent buffer to _maxLength? This feels like an inefficient way to do this -- i.e. to allocate a new array (the buffer in MemoryStream), write into it, and then allocate another copy of the array (in ToArray). I'm wondering if we can do it more efficiently?

Moving it here so it doesn't get lost

Reply:
Yes, it's just truncating the stream to max length. Maybe it's possible to use TryGetBuffer or do what you suggested in another comment, copying directly into a byte[]. I'll experiment a bit. I left it as-is from Azure.Core on the first iteration to get all the tests green but I think there are definitely some improvements that can be made.

Reply:
I poked around a bit and I think you can do it in a way that allocates the buffer only once. 🙂

I also think we might be able to use MemoryStream instead of MaxLengthStream if you think removing the custom type implementation would simplify the code complexity a bit.